Code/GraphMol/Substruct/SubstructMatch.cpp

```diff
@@ -130,16 +130,59 @@ void MatchSubqueries(const ROMol &mol, QueryAtom::QUERYATOM_QUERY *q,
                      SUBQUERY_MAP &subqueryMap,
                      std::vector<RecursiveStructureQuery *> &locked);
 
-bool matchCompare(const std::pair<int, int> &a, const std::pair<int, int> &b);
-bool matchVectCompare(const MatchVectType &a, const MatchVectType &b);
-bool isToBeAddedToVector(std::vector<MatchVectType> &matches,
-                         const MatchVectType &m);
-void mergeMatchVect(std::vector<MatchVectType> &matches,
-                    const std::vector<MatchVectType> &matchesTmp,
-                    const ResSubstructMatchHelperArgs_ &args);
+inline int getSecondPairElement(const std::pair<int, int> &p) {
+  return p.second;
+}
+
+bool insertIfNeeded(std::set<MatchVectType> &matches, const MatchVectType &m) {
+  bool shouldInsert = true;
+  std::unordered_set<int> matchAsSet;
+  std::transform(m.begin(), m.end(),
+                 std::inserter(matchAsSet, matchAsSet.begin()),
+                 getSecondPairElement);
+  for (auto it = matches.begin(); it != matches.end(); ++it) {
+    std::unordered_set<int> existingMatchAsSet;
+    std::transform(
+        it->begin(), it->end(),
+        std::inserter(existingMatchAsSet, existingMatchAsSet.begin()),
+        getSecondPairElement);
+    if (matchAsSet == existingMatchAsSet) {
+      if (m < *it) {
+        matches.erase(it);
+      } else {
+        shouldInsert = false;
+      }
+      break;
+    }
+  }
+  if (shouldInsert) {
+    matches.insert(m);
+  }
+  return shouldInsert;
+}
+
+template <
+    typename Container,
+    typename std::enable_if<
```

Comment:
Is this complication necessary?
It's cute to use language features, but isn't there a simpler way to express this?
(Note, there may not be... I haven't banged my head against the syntax and the reference material in order to figure out what this does yet)